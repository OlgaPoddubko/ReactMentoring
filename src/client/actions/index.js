export const LOAD_GALLERY = 'gallery/LOAD_GALLERY';
export const UPDATE_MOVIE_BY_ID = 'gallery/UPDATE_MOVIE_BY_ID';
export const UPDATE_GALLERY = 'gallery/UPDATE_GALLERY';
export const UPDATE_RELATED_MOVIES = 'gallery/UPDATE_RELATED_MOVIES';
export const CHANGE_SEARCH_INPUT = 'gallery/CHANGE_SEARCH_INPUT';
export const CHANGE_SEARCH_BY = 'gallery/CHANGE_SEARCH_BY';
export const CHANGE_SORT_BY = 'gallery/CHANGE_SORT_BY';
export const SET_CURRENT_MOVIE = 'gallery/SET_CURRENT_MOVIE';

export const loadGallery = () => ({
  type: LOAD_GALLERY,
});

export const updateMovieById = id => ({
  type: UPDATE_MOVIE_BY_ID,
  id,
});

export const updateGallery = gallery => ({
  type: UPDATE_GALLERY,
  gallery,
});

export const updateRelatedMovies = relatedMovies => ({
  type: UPDATE_RELATED_MOVIES,
  relatedMovies,
});

export const changeSearchInput = newSearchInput => ({
  type: CHANGE_SEARCH_INPUT,
  newSearchInput,
});

export const changeSearchBy = newSearchBy => ({
  type: CHANGE_SEARCH_BY,
  newSearchBy,
});

export const changeSortBy = newSortBy => ({
  type: CHANGE_SORT_BY,
  newSortBy,
});

export const setCurrentMovie = newCurrentMovie => ({
  type: SET_CURRENT_MOVIE,
  newCurrentMovie
})

export const fetchGallery = ( state = INITIAL_STATE ) => (dispatch) => {
  dispatch(loadGallery());
  let url = `http://react-cdp-api.herokuapp.com/movies?sortBy=${state.sortBy}&sortOrder=desc&search=${state.search}&searchBy=${state.searchBy}&limit=12`;
  return fetch(url)
    .then(res => res.json())
    .then(gallery => dispatch(updateGallery(gallery)));
};

export const fetchMovie = ( id ) => (dispatch) => {
  let url = `http://react-cdp-api.herokuapp.com/movies/${id}`;
  return fetch(url)
    .then(res => res.json())
    .then(movie => dispatch(setCurrentMovie(movie)));
};

export const fetchRelatedMovies = ( genre ) => (dispatch) => {
  dispatch(loadGallery());
  let url = `http://react-cdp-api.herokuapp.com/movies?search=${genre}&searchBy=genres&limit=12`;
  return fetch(url)
    .then(res => res.json())
    .then(relatedMovies => dispatch(updateRelatedMovies(relatedMovies)));
};
