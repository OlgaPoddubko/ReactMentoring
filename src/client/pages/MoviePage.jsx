import React from 'react';
import movieArr from '../movieArr';

import ErrorBoundary from '../components/ErrorBoundary';
import MovieInfo from '../components/MovieInfo';
import ProposeByGenre from '../components/ProposeByGenre';
import MovieList from '../components/MovieList';
import Footer from '../components/Footer';

class MoviePage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      data: movieArr.data,
      total: movieArr.total
    };
  }

  render() {
    return (
      <React.Fragment>
        <ErrorBoundary >
          <MovieInfo movie={movieArr.data[0]} />
          <ProposeByGenre movie={movieArr.data[0]} />
          <MovieList movies={this.state.data}/>
          <Footer />
        </ErrorBoundary>
      </React.Fragment>
    );
  }
}

export default MoviePage;
