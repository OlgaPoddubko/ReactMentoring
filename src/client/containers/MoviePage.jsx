// @flow
import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { withRouter } from 'react-router';

import { fetchMovie, fetchRelatedMovies } from '../actions';

import ErrorBoundary from '../components/ErrorBoundary';
import MovieInfo from '../components/MovieInfo';
import ProposeByGenre from '../components/ProposeByGenre';
import RelatedMovies from '../components/RelatedMovies';
import Footer from '../components/Footer';

type Props = {
  fetchMovie: Function,
  fetchRelatedMovies: Function,
  currentMovie: {
    genres?: Array<string>
  },
  match: {
   params: {
     movieId: string
   }
 },
}

class MoviePage extends React.Component<Props> {
  static fetchData(dispatch, match) {
      return dispatch(fetchMovie(match.params.movieId));
    }

  componentWillMount() {
    this.props.fetchMovie(this.props.match.params.movieId);
    if (this.props.currentMovie.genres) { this.props.fetchRelatedMovies(this.props.currentMovie.genres[0]); }
  }

  render() {
    return (
        <React.Fragment>
          <ErrorBoundary >
            <MovieInfo />
            <ProposeByGenre />
            <RelatedMovies />
            <Footer />
          </ErrorBoundary>
        </React.Fragment>
    );
  }
}

const mapStateToProps = ( state, {match} ) => ({
  currentMovie: state.gallery.currentMovie,
});

const mapDispatchToProps = dispatch => bindActionCreators({
  fetchMovie, fetchRelatedMovies
}, dispatch);

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(MoviePage));
