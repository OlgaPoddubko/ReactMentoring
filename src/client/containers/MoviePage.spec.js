import React from 'react';
import { MoviePage } from './MoviePage';
import { fetchMovie, fetchRelatedMovies } from '../actions';
import moviesArr from '../../../test/moviesArr';

describe('render MoviePage', () => {
  it('should render MoviePage', () => {
    const wrapper = shallow(<MoviePage
      currentMovie = { moviesArr.data[0] }
      match = {{ params: { movieId: '637' } }}
      fetchMovie = { fetchMovie }
      fetchRelatedMovies = { fetchRelatedMovies }
    />);
    expect(wrapper).toMatchSnapshot();
  });
});
