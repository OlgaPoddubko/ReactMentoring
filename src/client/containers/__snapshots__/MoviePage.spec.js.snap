// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render MoviePage should render MoviePage 1`] = `
<React.Fragment>
  <ErrorBoundary>
    <Connect(MovieInfo) />
    <Connect(ProposeByGenre) />
    <Connect(RelatedMovies) />
    <Footer />
  </ErrorBoundary>
</React.Fragment>
`;
